pkg/app/ops/apikeylastusedtimeupdater/apikeylastusedtimeupdater.go

```
@@ -0,0 +1,130 @@
+) *APIKeyLastUsedTimeUpdater {
+	return &APIKeyLastUsedTimeUpdater{
+		apiKeyStore:             datastore.NewAPIKeyStore(ds, datastore.OpsCommander),
+		apiKeyLastUsedTimeCache: rediscache.NewHashCache(rd, apiKeyLastUsedCacheHashKey),
+		logger:                  logger.Named("api-key-last-used-time-updater"),
+	}
+}
+
+func (c *APIKeyLastUsedTimeUpdater) Run(ctx context.Context) error {
+	c.logger.Info("start running APIKeyLastUsedTimeUpdater")
+
+	t := time.NewTicker(interval)
+	for {
+		select {
+		case <-ctx.Done():
+			c.logger.Info("APIKeyLastUsedTimeUpdater has been stopped")
+			return nil
+
+		case <-t.C:
+			start := time.Now()
+			if err := c.updateAPIKeyLastUsedTime(ctx); err == nil {
+				c.logger.Info("successfully update api key last used time", zap.Duration("duration", time.Since(start)))
+			}
+		}
+	}
+}
+
+func (c *APIKeyLastUsedTimeUpdater) updateAPIKeyLastUsedTime(ctx context.Context) error {
+
+	opts := datastore.ListOptions{
+		Filters: []datastore.ListFilter{
+			{
+				Field:    "Disabled",
+				Operator: datastore.OperatorEqual,
+				Value:    false,
+			},
+		},
+	}
+
+	apiKeys, err := c.apiKeyStore.List(ctx, opts)
+	if err != nil {
+		c.logger.Error("failed to list API key", zap.Error(err))
+		return err
+	}
+
+	for _, apiKey := range apiKeys {
+		lastUsedTimeByte, err := c.apiKeyLastUsedTimeCache.Get(apiKey.Id)
+		if err != nil {
+			c.logger.Error("failed to update last used time",
+				zap.String("id", apiKey.Id),
+				zap.Error(err),
+			)
+			continue
+		}
+
+		lastUsedTime := bytes2int64(lastUsedTimeByte.([]byte))
+		if err := c.apiKeyStore.UpdateLastUsedAt(ctx, apiKey.Id, lastUsedTime); err != nil {
+			c.logger.Error("failed to update last used time",
+				zap.String("id", apiKey.Id),
+				zap.Error(err),
+			)
+			continue
+		}
+	}
+
+	return nil
+}
+
+func bytes2int64(bytes []byte) int64 {
+	var numString string
+	for i := range bytes {
+		numString += string(bytes[i])
+	}
+	num, _ := strconv.ParseInt(numString, 10, 64)
```

not sure why do we ignore the error here 🤔